src/python/pants/backend/python/util_rules/local_dists.py

```diff
@@ -0,0 +1,168 @@
+        sources: PythonSourceFiles = PythonSourceFiles(
+            SourceFiles(EMPTY_SNAPSHOT, tuple()), tuple()
+        ),
+    ) -> None:
+        self.addresses = Addresses(addresses)
+        self.interpreter_constraints = interpreter_constraints
+        self.sources = sources
+
+
+@dataclass(frozen=True)
+class LocalDistsPex:
+    """A PEX file containing locally-built dists.
+
+    Can be consumed from another PEX, e.g., by adding to PEX_PATH.
+
+    Lists the files provided by the dists on sys.path, so they can be subtracted from
+    sources digests, to prevent the same file ending up on sys.path twice.
+    """
+
+    pex: Pex
+    # The sources from the request, but with any files provided by the local dists subtracted out.
+    remaining_sources: PythonSourceFiles
+
+
+@rule(desc="Building local distributions")
+async def build_local_dists(
+    request: LocalDistsPexRequest,
+) -> LocalDistsPex:
+
+    transitive_targets = await Get(TransitiveTargets, TransitiveTargetsRequest(request.addresses))
+
+    python_dist_field_sets = [
+        PythonDistributionFieldSet.create(target)
+        for target in transitive_targets.closure
+        if PythonDistributionFieldSet.is_applicable(target)
+    ]
+
+    dists = await MultiGet(
+        [Get(BuiltPackage, PackageFieldSet, field_set) for field_set in python_dist_field_sets]
+    )
+
+    # The primary use-case of the "local dists" feature is to support consuming native extensions
+    # as wheels without having to publish them first.
+    # It doesn't seem very useful to consume locally-built sdists, and it makes it hard to
+    # reason about possible sys.path collisions between the in-repo sources and whatever the
+    # sdist will place on the sys.path when it's installed.
+    # So for now we simply ignore sdists, with a warning if necessary.
+    provided_files = set()
+    wheels = []
+
+    for dist in dists:
+        contents = await Get(DigestContents, Digest, dist.digest)
+        artifacts = set((a.relpath or "") for a in dist.artifacts)
+        # A given local dist might build a wheel and an sdist (and maybe other artifacts -
+        # we don't know what setup command was run...)
+        # As long as there is a wheel, we can ignore the other artifacts.
+        wheel = next((art for art in artifacts if art.endswith(".whl")), None)
+        if wheel:
+            wheel_content = next(content for content in contents if content.path == wheel)
+            wheels.append(wheel)
+            buf = BytesIO()
+            buf.write(wheel_content.content)
+            buf.seek(0)
+            with zipfile.ZipFile(buf) as zf:
+                provided_files.update(zf.namelist())
+        else:
+            for art in artifacts:
+                logger.warning(
+                    f"Omitting dependency on local non-wheel artifact {art}. "
+                    f"This code will be depended on directly from sources."
```

Comment:
This warning misses context for why Pants was trying to do this in the first place. Given how huge the dep graph can be, it's not very intuitive imo.

How about including the root's address in `LocalDistsPexRequest`, so you can say something like:

> The target `{address}` transitively depends on a the `python_distribution` target `{dist_field_set.address}`, but it does not say to build a wheel via the `setup_py_commands` field. So, omitting a dependency on its wheel and instead depending on the code directly from sources. If you'd like to instead depend on the wheel, update the `python_distribition` - see {docs_url}.

Ack that that means `LocalDistsPexRequest` expects a single root target. That seems fine for right now. If that doesn't scale for future callers, we can revisit.

Reply:
That seems too elaborate to me. There definitely may be multiple roots even in trivial use cases, such as `./pants test ::`. But I can rephrase this in terms of the target instead of the artifact.

Reply:
Note though that `./pants test ::` calls the rule once-per-file. There _is_ always one root there in the current context.

I think `./pants repl ::` breaks that though.